diff -urN gcc-2.95.3/gcc/config/s390/s390.c gcc-2.95.3-s390/gcc/config/s390/s390.c
--- gcc-2.95.3/gcc/config/s390/s390.c	Wed Sep 12 13:20:15 2001
+++ gcc-2.95.3-s390/gcc/config/s390/s390.c	Mon Aug 20 14:52:25 2001
@@ -2319,8 +2319,7 @@
 
       /* Decrement stack.  */
 
-      if (TARGET_BACKCHAIN || (STARTING_FRAME_OFFSET +
-			       lsize + STACK_POINTER_OFFSET > 4095
+      if (TARGET_BACKCHAIN || (frame_size + STACK_POINTER_OFFSET > 4095
 			       || frame_pointer_needed
 			       || current_function_calls_alloca))
 	{
@@ -2360,8 +2359,7 @@
 
       /* Generate backchain.  */
 
-      if (TARGET_BACKCHAIN || (STARTING_FRAME_OFFSET + 
-			       lsize + STACK_POINTER_OFFSET > 4095
+      if (TARGET_BACKCHAIN || (frame_size + STACK_POINTER_OFFSET > 4095
 			       || frame_pointer_needed
 			       || current_function_calls_alloca))
 	{
diff -urN gcc-2.95.3/gcc/cse.c gcc-2.95.3-s390/gcc/cse.c
--- gcc-2.95.3/gcc/cse.c	Thu Jan 25 15:03:03 2001
+++ gcc-2.95.3-s390/gcc/cse.c	Mon Aug 20 14:52:55 2001
@@ -6578,8 +6578,16 @@
     }
   else if (GET_CODE (x) == CLOBBER)
     {
-      if (GET_CODE (XEXP (x, 0)) == MEM)
-	canon_reg (XEXP (x, 0), NULL_RTX);
+      rtx clobbered = XEXP (x, 0);
+
+      if (GET_CODE (clobbered) == MEM)
+	canon_reg (clobbered, NULL_RTX);
+      else if (GET_CODE (clobbered) == REG
+	       || GET_CODE (clobbered) == SUBREG)
+	invalidate (clobbered, VOIDmode);
+      else if (GET_CODE (clobbered) == STRICT_LOW_PART
+	       || GET_CODE (clobbered) == ZERO_EXTRACT)
+	invalidate (XEXP (clobbered, 0), GET_MODE (clobbered));
     }
 
   /* Canonicalize a USE of a pseudo register or memory location.  */
diff -urN gcc-2.95.3/gcc/f/com.h gcc-2.95.3-s390/gcc/f/com.h
--- gcc-2.95.3/gcc/f/com.h	Sat Apr 17 12:58:25 1999
+++ gcc-2.95.3-s390/gcc/f/com.h	Fri Sep  7 17:58:25 2001
@@ -127,7 +127,7 @@
 #define LONG_DOUBLE_TYPE_SIZE (BITS_PER_WORD * 2)
 #endif
 
-#if LONG_TYPE_SIZE == FLOAT_TYPE_SIZE
+#if (LONG_TYPE_SIZE == MAX_LONG_TYPE_SIZE) && (LONG_TYPE_SIZE == FLOAT_TYPE_SIZE)
 #  define FFECOM_f2cINTEGER FFECOM_f2ccodeLONG
 #  define FFECOM_f2cLOGICAL FFECOM_f2ccodeLONG
 #elif INT_TYPE_SIZE == FLOAT_TYPE_SIZE
@@ -137,7 +137,7 @@
 #  error Cannot find a suitable type for FFECOM_f2cINTEGER
 #endif
 
-#if LONG_TYPE_SIZE == (FLOAT_TYPE_SIZE * 2)
+#if (LONG_TYPE_SIZE == MAX_LOG_TYPE_SIZE) && (LONG_TYPE_SIZE == (FLOAT_TYPE_SIZE * 2))
 #  define FFECOM_f2cLONGINT FFECOM_f2ccodeLONG
 #elif LONG_LONG_TYPE_SIZE == (FLOAT_TYPE_SIZE * 2)
 #  define FFECOM_f2cLONGINT FFECOM_f2ccodeLONGLONG
